packages/jupyter-ai/src/components/chat-settings.tsx

```diff
@@ -16,234 +16,164 @@ import {
-    );
-    const emp = emProviders?.providers.find(
-      provider => provider.id === selectedEmpId
-    );
-    const newApiKeys: Record<string, string> = {};
+    if (server.state !== ServerInfoState.Ready) {
+      return;
+    }
 
-    if (lmp?.auth_strategy && lmp.auth_strategy.type === 'env') {
-      newApiKeys[lmp.auth_strategy.name] =
-        config?.api_keys[lmp.auth_strategy.name] || '';
+    const newApiKeys: Record<string, string> = {};
+    const lmAuth = lmProvider?.auth_strategy;
+    const emAuth = emProvider?.auth_strategy;
+    if (
+      lmAuth?.type === 'env' &&
+      !server.config.api_keys.includes(lmAuth.name)
+    ) {
+      newApiKeys[lmAuth.name] = '';
     }
-    if (emp?.auth_strategy && emp.auth_strategy.type === 'env') {
-      newApiKeys[emp.auth_strategy.name] =
-        config?.api_keys[emp.auth_strategy.name] || '';
+    if (
+      emAuth?.type === 'env' &&
+      !server.config.api_keys.includes(emAuth.name)
+    ) {
+      newApiKeys[emAuth.name] = '';
     }
 
-    setInputConfig(inputConfig => ({
-      ...inputConfig,
-      api_keys: { ...config?.api_keys, ...newApiKeys }
-    }));
-  }, [inputConfig.model_provider_id, inputConfig.embeddings_provider_id]);
+    setApiKeys(newApiKeys);
+  }, [lmProvider, emProvider, server]);
 
-  const handleSave = async () => {
-    const inputConfigCopy: AiService.Config = {
-      ...inputConfig,
-      model_provider_id: lmGlobalId,
-      api_keys: { ...inputConfig.api_keys },
-      send_with_shift_enter: inputConfig.send_with_shift_enter ?? true
-    };
-
-    // delete any empty api keys
-    for (const apiKey in inputConfigCopy.api_keys) {
-      if (inputConfigCopy.api_keys[apiKey] === '') {
-        delete inputConfigCopy.api_keys[apiKey];
-      }
+  /**
+   * Effect: re-initialize fields object whenever the selected LM changes.
+   */
+  useEffect(() => {
+    if (server.state !== ServerInfoState.Ready || !lmGlobalId) {
+      return;
     }
 
+    const currFields: Record<string, any> =
+      server.config.fields?.[lmGlobalId] ?? {};
+    setFields(currFields);
+  }, [server, lmProvider]);
+
+  const handleSave = async () => {
     // compress fields with JSON values
-    for (const gmid in inputConfigCopy.fields) {
-      for (const fieldKey in inputConfigCopy.fields[gmid]) {
-        const fieldVal = inputConfigCopy.fields[gmid][fieldKey];
-        if (typeof fieldVal !== 'string') {
-          continue;
-        }
+    if (server.state !== ServerInfoState.Ready) {
+      return;
+    }
 
-        try {
-          const parsedFieldVal = JSON.parse(fieldVal);
-          const compressedFieldVal = JSON.stringify(parsedFieldVal);
-          inputConfigCopy.fields[gmid][fieldKey] = compressedFieldVal;
-        } catch (e) {
-          continue;
-        }
+    for (const fieldKey in fields) {
+      const fieldVal = fields[fieldKey];
+      if (typeof fieldVal !== 'string' || !fieldVal.trim().startsWith('{')) {
```

Comment:
Technically, not all JSON values need to start with a `{`. Why not parse with `JSON.parse` or another parser, as we did before?

Reply:
Right now, the only place this is used is for SM Endpoints, where users have to provide a JSON request schema, which has to be an object for now. We will probably have to rework this logic later when we have more complex model fields.